initial support of slack socket mode - refactor

ryancurrah opened the pull request:
I played around with the Slack integration what I was looking for is to use socket mode so we don't have to expose this chat bot to the internet. Here is my WIP.

jaimeteb commented:
Thank you @ryancurrah ! Allow me to move things around a little bit, but it looks promising!

ryancurrah commented:
I have another rather large commit to push. I changed the channels up a bit.

ryancurrah commented:
Ok pushed it. It's not working because I working on changing the messages before stopping

jaimeteb commented:
> Ok pushed it. It's not working because I was working on changing the messages before stopping

OK, yes, I noticed the missing `messages` package. I'm liking the refactor a lot, I'll wait for the `messages` change.

ryancurrah commented:
I'll have an update tonight.

ryancurrah commented:
I havent forgot about this. Work has been super busy. I might have to finish this up on the weekend.

jaimeteb commented:
No worries, it's been busy for me as well. Thank you !

ryancurrah commented:
Just pushed another commit. Still working on this. I'm going to comment on a bunch of changes once done.

jaimeteb commented:
> Just pushed another commit. Still working on this. I'm going to comment on a bunch of changes once done.

This is awesome, thank you for these changes, you've inspired me to refactor the `clf` package before adding more classifier algorithms. Looking forward to the changes and comments Ryan!

jaimeteb commented:
I updated the examples' extensions with the new structure and added some helper functions in the `query` package

ryancurrah commented:
Ok I think I am done with the refactoring. I just need to fix the tests.

ryancurrah commented on .golangci.yml:
```diff
@@ -0,0 +1,115 @@
+run:
```
Added a `golangci-lint` config other contributors could use to ensure some quality.

ryancurrah commented on .goreleaser.yml:
```diff
@@ -0,0 +1,30 @@
+# This is an example goreleaser.yaml file with some sane defaults.
```
Added a `goreleaser` config to make publishing releases easier.

ryancurrah commented on .travis.yml:
```diff
@@ -12,9 +12,9 @@ before_script:
- - go test ./fsm ./clf ./bot ./ext -cover -coverprofile=coverage.txt
+  - go run examples/00_test/ext/ext.go &
+  - go run examples/03_pokemon/ext/ext.go -port 6770 &
+  - go test ./... -cover -coverprofile=coverage.txt
```
Adding `./...` here ensures we run all the tests. Not sure if you knew about this.

ryancurrah commented on Makefile:
```diff
@@ -1 +1,4 @@
 target: ;
+
+test:
+	go test ./... -cover -coverprofile=coverage.txt
```
Convenience make target.

ryancurrah replied:
Done.

ryancurrah commented on bot/bot.go:
```diff
@@ -5,24 +5,25 @@ import (
 	Name       string
 	Machines   fsm.StoreFSM
-	Domain     fsm.Domain
+	DB         *fsm.DB
```
`Domain` is a generic term and does not describe what it provides. I though maybe `DB` made sense here but that's debatable.

jaimeteb replied:
I thought about Domain at first because [Rasa calls it that](https://rasa.com/docs/rasa/domain/). In the future, I plan to add an actual database to store the conversation history, so `DB` may be confusing. But until then, I'll stick with it.

ryancurrah replied:
I did not know that. I will revert this back to `Domain`. Sorry about that I got carried away here.

jaimeteb replied:
No need to apologize!

ryancurrah commented on bot/bot.go:
```diff
@@ -5,24 +5,25 @@ import (
-	Extension  ext.Extension
-	Clients    Clients
+	Extension  extension.Extension
+	Channels   *channels.Channels
```
Renamed `Clients` to `Channels` because that's what we are using. Seems to make more sense.

ryancurrah commented on bot/bot.go:
```diff
@@ -31,34 +32,40 @@ type Prediction struct {
 // Answer takes a user input and executes a transition on the FSM if possible
-func (b Bot) Answer(mess cmn.Message) interface{} {
-	if !b.Machines.Exists(mess.Sender) {
+func (b *Bot) Answer(question *query.Question) ([]query.Answer, error) {
```
Now returns an error if processing an answer fails.

ryancurrah commented on bot/bot.go:
```diff
@@ -92,28 +102,36 @@ func LoadName(bcName string) (name string) {
 
 // LoadBot loads all configurations and returns a Bot
-func LoadBot(path *string) Bot {
+func LoadBot(path *string) (*Bot, error) {
```
Now returns an error if loading the config failed.

ryancurrah commented on bot/bot_test.go:
```diff
@@ -64,74 +88,100 @@ func TestBot2(t *testing.T) {
-
-	SendMessages(new(interface{}), &bot.Clients.REST, "8809", wREST)
-}
+// func TestBotNoClientsAndImages(t *testing.T) {
```
- [ ]  Need to get this test working still.

ryancurrah replied:
Done.

ryancurrah commented on bot/cli.go:
```diff
@@ -11,30 +11,40 @@ import (
 // SendAndReceive send a message to localhost endpoint and receives an answer
-func SendAndReceive(mess *cmn.Message, url string) *[]cmn.Message {
-	jsonMess, _ := json.Marshal(mess)
+func SendAndReceive(question *query.Question, url string) []query.Answer {
```
A slice is a pointer already so no need to say return a pointer `*`

ryancurrah commented on bot/server.go:
```diff
@@ -5,119 +5,145 @@ import (
+	b.endpointHandler(w, r, b.Channels.Slack)
+}
+
+func (b *Bot) endpointHandler(w http.ResponseWriter, r *http.Request, chnl channels.Channel) {
```
Because we are now properly using the `Channel` interface we can generalize the endpoint handler for all our clients and reduce duplicate code.

ryancurrah commented on bot/server.go:
```diff
@@ -5,119 +5,145 @@ import (
-		log.Error(err)
-		return
-	}
+func (b *Bot) slackMessageEvents() {
```
Here we handle the events we receive from Slack.

ryancurrah commented on bot/server.go:
```diff
@@ -5,119 +5,145 @@ import (
 	log.Info("Server started")
 
+	// Event listeners
+	bot.slackMessageEvents()
```
Start the Slack event listener.

ryancurrah commented on bot/server.go:
```diff
@@ -130,5 +156,21 @@ func ServeBot(path *string, port *int) {
+	log.Fatal(http.ListenAndServe(fmt.Sprintf(":%d", *port), r))
+}
+
+func writeAnswer(w http.ResponseWriter, answers []query.Answer) {
```
This was pulled out of the `Channel` clients so the channel implementation can be a little more generic.

ryancurrah commented on channels/channels.go:
```diff
@@ -0,0 +1,87 @@
+package channels
```
Moved `channels` into it's own package because it is a huge package. I can see this growing even more making it harder to move in the future, so lets do it now.

ryancurrah commented on channels/channels.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// Channels combines all available channel clients
+type Channels struct {
```
Declare all our `Channels` here and ensure they conform to the `Channel` interface. This allowed us to reduce duplicate code in the `bot` package.

ryancurrah commented on channels/channels.go:
```diff
@@ -0,0 +1,87 @@
+	// ReceiveMessage from the channel
+	ReceiveMessage(w http.ResponseWriter, r *http.Request) (*messages.Receive, error)
+	// ReceiveMessages from the channel. Starts a long running process, receives questions and sends them to the receiveChan
+	ReceiveMessages(receiveChan chan messages.Receive)
```
Added a new function `ReceiveMessages` which is meant to be a long running process. I think the name is ok but fine with changing it. It takes a channel which it will send messages to the bot on. The bot will process the messages.

jaimeteb replied:
Thank you! This is ideal for socket connections!

ryancurrah commented on channels/messages/messages.go:
```diff
@@ -0,0 +1,38 @@
+package messages
```
This is a new package in the `channels` package that allows us to do a couple things. 

1. Have a well defined receive and response message structure by which all channels can use.
2. Allows to control the reply options for each channel. Every channel will be a little different.

ryancurrah commented on channels/slack/slack.go:
```diff
@@ -0,0 +1,246 @@
+}
+
+// ReceiveMessages uses event queues to receive messages. Starts a long running process
+func (c *Channel) ReceiveMessages(receiveChan chan messages.Receive) {
```
This is the meat of my Slack change. This allows the chat bot to receive messages but not be exposed to the internet. Super cool.

ryancurrah commented on clf/nb.go:
```diff
@@ -70,21 +73,23 @@ func Create(path *string) Classifier {
-			classifier.Learn(Pipeline(&txt, &pipeline), bayesian.Class(cls.Command))
+	for n := range cls {
+		for i := range cls[n].Texts {
+			classifier.Learn(Pipeline(&cls[n].Texts[i], &pipeline), bayesian.Class(cls[n].Command))
```
This was changed because you cannot pass a pointer of a variable that exists only in the for loop. The variables that get created by for loops only exist in the loop. After that they get destroyed. The solution is to pass by index.

jaimeteb replied:
I come from a Python background, sometimes I have trouble with pointers 😅, thank you for this change and the explanation!

ryancurrah commented on clf/pipeline.go:
```diff
@@ -5,6 +5,8 @@ import (
 	"strings"
 )
 
+var removeSymbolRe = regexp.MustCompile(`\W`)
```
Another linter issue here. Best practise is to use `MustCompile`. This does the regex compile at build time thus saving processing power.

ryancurrah commented on fsm/config.go:
```diff
@@ -0,0 +1,118 @@
+package fsm
```
I split out the `config` type things from the `fsm.go` file. It was getting large.

ryancurrah commented on logger/logger.go:
```diff
@@ -0,0 +1,27 @@
+		log.SetLevel(log.InfoLevel)
+	}
+
+	log.SetFormatter(&runtime.Formatter{
```
Ok so feel free to ask me to revert this. I added a plugin to the logger that enabled reporting the file. line number and function that a log event occurred on. It made it easier to troubleshoot issues.

```
time="2021-02-09 23:18:21" level=info msg="Added Slack client: xoxb-63923..." file=/Users/ryancurrah/git/chatto/channels/slack/slack.go function=NewChannel line=50
time="2021-02-09 23:18:21" level=info msg="Registered CacheStoreFSM" file=/Users/ryancurrah/git/chatto/fsm/store.go function=LoadStore line=161
time="2021-02-09 23:18:21" level=info msg="* TTL:    -1" file=/Users/ryancurrah/git/chatto/fsm/store.go function=LoadStore line=162
time="2021-02-09 23:18:21" level=info msg="* Purge:  -1" file=/Users/ryancurrah/git/chatto/fsm/store.go function=LoadStore line=163
time="2021-02-09 23:18:21" level=info msg="Server started" file=/Users/ryancurrah/git/chatto/bot/server.go function=ServeBot line=142
time="2021-02-09 23:18:21" level=info msg="Connecting to Slack with Socket Mode..." file=/Users/ryancurrah/git/chatto/channels/slack/slack.go function=func1 line=168
time="2021-02-09 23:18:23" level=info msg="Connected to Slack with Socket Mode" file=/Users/ryancurrah/git/chatto/channels/slack/slack.go function=func1 line=172
```

jaimeteb replied:
How about enabling the `runtime.Formatter` only when `DEBUG` is set to `true`?
That way, in a non-debugging environment, the logs are cleaner.
Great addition, by the way.

ryancurrah replied:
Good call. I will do that.

ryancurrah commented on fsm/config.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// Function lists the transitions available for the FSM
+type Function struct {
```
I added some controversial changes here that break things.

ryancurrah commented on fsm/config.go:
```diff
@@ -0,0 +1,118 @@
+	Transition Transition `yaml:"transition"`
+	Command    string     `yaml:"command"`
+	Slot       Slot       `yaml:"slot"`
+	Extension  string     `yaml:"extension"`
```
Instead of specifying `ext_foo` in the message I added a dedicated `Extension` field. Where one would choose to use an extension instead of message.

jaimeteb replied:
Was thinking about making this change eventually, thank you for anticipating it!

ryancurrah commented on fsm/config.go:
```diff
@@ -0,0 +1,118 @@
+	Command    string     `yaml:"command"`
+	Slot       Slot       `yaml:"slot"`
+	Extension  string     `yaml:"extension"`
+	Message    []Message  `yaml:"message"`
```
Message is now well defined. You cannot mix and match formats anymore, it has to be...

```
message:
  - text: "foo"
  - text: "bar"
  - image: "https://foo/bar/image"
```

I did this because passing around an interface everywhere made it super difficult to work with this. Feel free to raise concerns here.

jaimeteb replied:
Ok but you can still have an image with text, right? Like this:

```yaml
message:
  - text: "first message, only text"
  - text: "second message, with image"
    image: "https://foo/bar/image"
```

ryancurrah replied:
Yep that is still supported.

ryancurrah commented on fsm/fsm.go:
```diff
@@ -4,62 +4,33 @@ import (
+// DB contains BaseDB plus the functions required for a fully
+// functioning FSM
+type DB struct {
+	BaseDB
```
I saved us some code here by using composition. Seem to fit well here.

ryancurrah commented on fsm/fsm.go:
```diff
@@ -69,35 +40,25 @@ type CmdStateTuple struct {
 
 // TransitionFunc models a transition function
-type TransitionFunc func(m *FSM) interface{}
+type TransitionFunc func(m *FSM) (string, []Message)
```
Due to the change of the fsm `Function` we no longer have to pass around an interface as a value.

ryancurrah commented:
I've fixed most of the tests. Still got a 3 or 4 to fix up, I'll try and get to them tomorrow.

jaimeteb commented:
> I've fixed most of the tests. Still got a 3 or 4 to fix up, I'll try and get to them tomorrow.

Thank you so much, this is looking very good and way better organized. I'll start updating the docs.

jaimeteb commented on bot/server.go:
```diff
@@ -58,25 +58,27 @@ func (b *Bot) endpointHandler(w http.ResponseWriter, r *http.Request, chnl chann
-			if err != nil {
-				log.Error(err)
-				return
+	if b.Channels.Slack != nil {
```
I added this condition to skip this process if there is no Slack channel configuration

jaimeteb commented on channels/channels.go:
```diff
@@ -49,20 +49,23 @@ func Load(path *string) *Channels {
 		switch err.(type) {
 		case viper.ConfigFileNotFoundError:
-			log.Warn("File chn.yml not found, skipping channels")
+			log.Warn("File chn.yml not found, using only REST channel")
```
Created the REST channel before loading the channels, in order to use it as the default channel (to use it in cli mode for example).

jaimeteb commented on bot/bot.go:
```diff
@@ -59,7 +59,7 @@ func (b *Bot) Answer(question *query.Question) ([]query.Answer, error) {
 		reply, err = b.Extension.RunExtFunc(question, runExt, b.DB, machine)
 		if err != nil {
-			return nil, err
+			log.Error(err) // return nil, err
```
Since errors from `RunExtFunc` will be converted into error replies, I think the error returned from this `Answer` function can be deprecated. What do you think @ryancurrah ?

ryancurrah replied:
Done.

jaimeteb commented:
@ryancurrah I'm testing the Slack socket mode but I'm getting:

```
ERRO[2021-02-10 17:33:53] not_authed:                                   file=/mnt/c/Users/jaime.tenorio/Documents/projects/chatto/channels/slack/slack.go function=SendMessage line=81
```

I enabled socket mode in my app and copied the generated token, but, is there something else that has to be done? I'd like to fully understand this process in order to write the docs.

ryancurrah commented:
> @ryancurrah I'm testing the Slack socket mode but I'm getting:
> 
> ```
> ERRO[2021-02-10 17:33:53] not_authed:                                   file=/mnt/c/Users/jaime.tenorio/Documents/projects/chatto/channels/slack/slack.go function=SendMessage line=81
> ```
> 
> I enabled socket mode in my app and copied the generated token, but, is there something else that has to be done? I'd like to fully understand this process in order to write the docs.

Yeah it is not trivial to setup. I followed the guide here: https://api.slack.com/apis/connections/socket#toggling.

You need two tokens a regular token and an app token. Then you need to turn on socket mode, which the guide above mentions. And you need to scope the permissions properly.

Here are some screenshots of my setup.

<img width="721" alt="Screen Shot 2021-02-11 at 10 35 52 AM" src="https://user-images.githubusercontent.com/6209771/107659664-5e986e00-6c55-11eb-8cd4-494345c97045.png">
<img width="966" alt="Screen Shot 2021-02-11 at 10 37 03 AM" src="https://user-images.githubusercontent.com/6209771/107659667-5fc99b00-6c55-11eb-9368-9b4b5ee450a0.png">
<img width="993" alt="Screen Shot 2021-02-11 at 10 38 20 AM" src="https://user-images.githubusercontent.com/6209771/107659672-60623180-6c55-11eb-8e0c-4adb5f20ae41.png">

ryancurrah commented:
I should note the permissions above will allow `chatto` to listen to all channels it is added to not just direct messages.

jaimeteb commented:
@ryancurrah Thanks for the help with the socket mode, I thought I only needed the `app_token`, but including the `token` as well did the trick. I'll add this to the docs!

ryancurrah commented:
No problem. After reworking the tests I have another rather large PR coming. Working on it now actually.

jaimeteb commented:
> No problem. After reworking the tests I have another rather large commit coming. Working on it now actually.

Can't wait !

ryancurrah commented:
Ok @jaimeteb I think I am done, pending your review.

ryancurrah commented:
Things to note in the latest change...

- Added a bunch of mocks for mocking responses in unit tests.
- Renamed `examples` to `_examples` so tools like `go test`, `golangci-lint`, etc won't try to recognize them. Prefixing directories with `_` in Go is a special meaning Go will ignore that dir.
- Used more Go'ish naming like `Config`, `New`, etc.
- Removed `ext_` from all extension function names since it is not needed anymore.
- Split out config from `bot.go` into `config.go`.
- Simplified the `Channel` interface by passing `[]byte` instead of http reader and writer for `ReceiveMessage` function.
- Fix panic when trying to access a `nil` `mux.Vars`.
- Added `testutils` package that has some functions like...
    - Get next free port on host for usage in unit tests.
    - Start an `extension` for running unit tests. 
- Don't register routes if `Channels` is `nil` or a specific `Channel`  also `nil`.
- Add table driven tests for `bot` and `channels`.
- `Channel` clients are now defined by an `interface{}` so we can mock them.

ryancurrah commented:
Okay now I'm done, done.

jaimeteb commented:
Great! Let me check it out and do some local testing

jaimeteb commented:
Still checking the latest changes. In the meantime @ryancurrah , can you explain to me how and when to use the `goreleaser` tool? I see that you included configuration for building Docker images and binaries. I used to push the images with GitHub Actions and add binaries to releases manually, so how should it be done now with this tool?
Sorry if these are dumb questions, I'm relatively new to Go and open-source projects like this.

ryancurrah commented:
@jaimeteb no problem. So `goreleaser` does all the work for you in terms of building and publishing artifacts for your software. Including Github releases and publishing to Dockerhub. There is support for Github actions: https://github.com/goreleaser/goreleaser-action. You can replace what you are doing. You can also try it locally: https://goreleaser.com/install/. Once installed on my branch run `make snapshot`. You will notice a new `dist/` folder with a bunch artifacts in it. The `snapshot` will not publish just build.

jaimeteb commented:
> @jaimeteb no problem. So `goreleaser` does all the work for you in terms of building and publishing artifacts for your Go software. Including Github releases and publishing to Dockerhub. There is support for Github actions: https://github.com/goreleaser/goreleaser-action. You can replace what you are doing. You can also try it locally: https://goreleaser.com/install/. Once installed. Switch to my branch run `make snapshot`. You will notice a new `dist/` folder with a bunch artifacts in it. The `snapshot` will not publish just build.

Thank you Ryan, I didn't know about this amazing tool!

jaimeteb commented:
Ok, I think I'm done with my local tests, I'll update the docs and merge.

ryancurrah commented:
Cool. I joined your Slack workspace.

jaimeteb commented:
Thank you very much for your contribution @ryancurrah, the project is looking way better now! 🤝🤩
